Add recipe `pyreqwest_impersonate`

### Versions

- Python: 3.10
- OS: Linux mint
- Kivy: 2.2.1
- Cython: 0.29.33
- OpenJDK: 17

### Description

I made the recipe according to the example:
```
from pythonforandroid.recipe import RustCompiledComponentsRecipe


class Pyreqwest_impersonateRecipe(RustCompiledComponentsRecipe):
    version = "v0.4.1"
    url = "https://github.com/deedy5/pyreqwest_impersonate/archive/refs/tags/{version}.tar.gz"
    use_maturin = True
    hostpython_prerequisites = ["typing_extensions"]
    site_packages_name = "pyreqwest_impersonate"


recipe = Pyreqwest_impersonateRecipe()
```

### buildozer.spec

```
# (str) The directory in which python-for-android should look for your own build recipes (if any)
p4a.local_recipes = /home/odinson/PycharmProjects/OSINT/src/python-for-android/recipes/
```

### Logs

```
File "/home/odinson/PycharmProjects/OSINT/src/python-for-android/recipes/pyreqwest_impersonate/__init__.py", line 1, in <module>
    from pythonforandroid.recipe import RustCompiledComponentsRecipe
ImportError: cannot import name 'RustCompiledComponentsRecipe' from 'pythonforandroid.recipe' (/home/odinson/PycharmProjects/OSINT/.buildozer/android/platform/python-for-android/pythonforandroid/recipe.py)

```

`buildozer clean`  I did too. I did everything as required according to the documentation.

@T-Dynamos log buildozer
```
[100%] Built target crypto
  cargo:root=/home/odinson/PycharmProjects/OSINT/.buildozer/android/platform/build-arm64-v8a/build/other_builds/pyreqwest_impersonate/arm64-v8a__ndk_target_21/pyreqwest_impersonate/target/aarch64-linux-android/debug/build/boring-sys-imp-169f8b4fd7943b34/out
  cargo:rustc-link-search=native=/home/odinson/PycharmProjects/OSINT/.buildozer/android/platform/build-arm64-v8a/build/other_builds/pyreqwest_impersonate/arm64-v8a__ndk_target_21/pyreqwest_impersonate/target/aarch64-linux-android/debug/build/boring-sys-imp-169f8b4fd7943b34/out/build/
  cargo:rustc-link-lib=static=crypto
  cargo:rustc-link-lib=static=ssl
  cargo:rerun-if-env-changed=BORING_BSSL_INCLUDE_PATH

  --- stderr
  android arch=aarch64 add ANDROID_ABI=arm64-v8a
  android toolchain=/home/odinson/.buildozer/android/platform/android-ndk-r25b/build/cmake/android.toolchain.cmake
  make[1]: warning: -j4 forced in submake: resetting jobserver mode.
  make[2]: warning: -j4 forced in submake: resetting jobserver mode.
  make[3]: warning: -j4 forced in submake: resetting jobserver mode.
  make[3]: warning: -j4 forced in submake: resetting jobserver mode.
  make[3]: warning: -j4 forced in submake: resetting jobserver mode.
  make[3]: warning: -j4 forced in submake: resetting jobserver mode.
  make[1]: warning: -j4 forced in submake: resetting jobserver mode.
  make[2]: warning: -j4 forced in submake: resetting jobserver mode.
  make[3]: warning: -j4 forced in submake: resetting jobserver mode.
  make[3]: warning: -j4 forced in submake: resetting jobserver mode.
  deps/boringssl/src/include/openssl/base.h:61:10: fatal error: 'sys/types.h' file not found
  thread 'main' panicked at /home/odinson/.cargo/registry/src/index.crates.io-6f17d22bba15001f/boring-sys-imp-2.1.0/build.rs:558:39:
  Unable to generate bindings: ClangDiagnostic("deps/boringssl/src/include/openssl/base.h:61:10: fatal error: 'sys/types.h' file not found\n")
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
💥 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit status: 101": `env -u CARGO PYO3_ENVIRONMENT_SIGNATURE="cpython-3.11-64bit" PYO3_PYTHON="/home/odinson/PycharmProjects/OSINT/.buildozer/android/platform/build-arm64-v8a/build/other_builds/hostpython3/desktop/hostpython3/native-build/python3.11" PYTHON_SYS_EXECUTABLE="/home/odinson/PycharmProjects/OSINT/.buildozer/android/platform/build-arm64-v8a/build/other_builds/hostpython3/desktop/hostpython3/native-build/python3.11" "cargo" "rustc" "--features" "pyo3/extension-module" "--target" "aarch64-linux-android" "--message-format" "json-render-diagnostics" "--manifest-path" "/home/odinson/PycharmProjects/OSINT/.buildozer/android/platform/build-arm64-v8a/build/other_builds/pyreqwest_impersonate/arm64-v8a__ndk_target_21/pyreqwest_impersonate/Cargo.toml" "--lib"`
```

@deedy5 Maybe you can help here:

```
Compiling encoding_rs v0.8.34====>  ] 189/201: mime_guess, pyo3            
error[E0432]: unresolved import `simd_funcs`/201: encoding_rs, pyo3           
  --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/x_user_defined.rs:16:13
   |
16 |         use simd_funcs::*;
   |             ^^^^^^^^^^ use of undeclared crate or module `simd_funcs`

error[E0412]: cannot find type `u16x8` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:129:40
    |
129 |     pub fn simd_at(&self, i: usize) -> u16x8 {
    |                                        ^^^^^ not found in this scope
    |
help: consider importing this type alias
    |
39  + use core::simd::u16x8;
    |

error[E0425]: cannot find value `SIMD_STRIDE_SIZE` in this scope
    --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:131:21
     |
131  |         assert!(i + SIMD_STRIDE_SIZE / 2 <= self.len);
     |                     ^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `ALU_STRIDE_SIZE`
     |
    ::: /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/ascii.rs:1323:9
     |
1323 |         pub const ALU_STRIDE_SIZE: usize = 8;
     |         ------------------------------------- similarly named constant `ALU_STRIDE_SIZE` defined here

error[E0425]: cannot find value `SIMD_STRIDE_SIZE` in this scope
    --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:161:12
     |
161  |         if SIMD_STRIDE_SIZE / 2 <= self.len {
     |            ^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `ALU_STRIDE_SIZE`
     |
    ::: /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/ascii.rs:1323:9
     |
1323 |         pub const ALU_STRIDE_SIZE: usize = 8;
     |         ------------------------------------- similarly named constant `ALU_STRIDE_SIZE` defined here

error[E0425]: cannot find value `SIMD_STRIDE_SIZE` in this scope
    --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:162:47
     |
162  |             let len_minus_stride = self.len - SIMD_STRIDE_SIZE / 2;
     |                                               ^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `ALU_STRIDE_SIZE`
     |
    ::: /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/ascii.rs:1323:9
     |
1323 |         pub const ALU_STRIDE_SIZE: usize = 8;
     |         ------------------------------------- similarly named constant `ALU_STRIDE_SIZE` defined here

error[E0425]: cannot find value `SIMD_STRIDE_SIZE` in this scope
    --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:175:27
     |
175  |                 offset += SIMD_STRIDE_SIZE / 2;
     |                           ^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `ALU_STRIDE_SIZE`
     |
    ::: /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/ascii.rs:1323:9
     |
1323 |         pub const ALU_STRIDE_SIZE: usize = 8;
     |         ------------------------------------- similarly named constant `ALU_STRIDE_SIZE` defined here

error[E0425]: cannot find value `SIMD_STRIDE_SIZE` in this scope
    --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:256:8
     |
256  |     if SIMD_STRIDE_SIZE <= len {
     |        ^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `ALU_STRIDE_SIZE`
     |
    ::: /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/ascii.rs:1323:9
     |
1323 |         pub const ALU_STRIDE_SIZE: usize = 8;
     |         ------------------------------------- similarly named constant `ALU_STRIDE_SIZE` defined here

error[E0425]: cannot find value `SIMD_STRIDE_SIZE` in this scope
    --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:257:38
     |
257  |         let len_minus_stride = len - SIMD_STRIDE_SIZE;
     |                                      ^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `ALU_STRIDE_SIZE`
     |
    ::: /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/ascii.rs:1323:9
     |
1323 |         pub const ALU_STRIDE_SIZE: usize = 8;
     |         ------------------------------------- similarly named constant `ALU_STRIDE_SIZE` defined here

error[E0425]: cannot find value `SIMD_STRIDE_SIZE` in this scope
    --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:260:52
     |
260  | ...rc.simd_at(offset + (SIMD_STRIDE_SIZE / 2));
     |                         ^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `ALU_STRIDE_SIZE`
     |
    ::: /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/ascii.rs:1323:9
     |
1323 |         pub const ALU_STRIDE_SIZE: usize = 8;
     |         ------------------------------------- similarly named constant `ALU_STRIDE_SIZE` defined here

error[E0425]: cannot find value `SIMD_STRIDE_SIZE` in this scope
    --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:273:23
     |
273  |             offset += SIMD_STRIDE_SIZE;
     |                       ^^^^^^^^^^^^^^^^ help: a constant with a similar name exists: `ALU_STRIDE_SIZE`
     |
    ::: /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/ascii.rs:1323:9
     |
1323 |         pub const ALU_STRIDE_SIZE: usize = 8;
     |         ------------------------------------- similarly named constant `ALU_STRIDE_SIZE` defined here

error[E0425]: cannot find function `load16_unaligned` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/x_user_defined.rs:125:34
    |
125 |             let input = unsafe { load16_unaligned(src_ptr.add(i * 16)) };
    |                                  ^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `simd_unpack` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/x_user_defined.rs:126:35
    |
126 |             let (first, second) = simd_unpack(input);
    |                                   ^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `store8_unaligned` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/x_user_defined.rs:129:17
    |
129 |                 store8_unaligned(dst_ptr.add(i * 16), shift_upper(first));
    |                 ^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `store8_unaligned` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/x_user_defined.rs:130:17
    |
130 | ...   store8_unaligned(dst_ptr.add((i * 16) + 8), shift_upper(second));
    |       ^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `load16_unaligned` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:137:31
    |
137 | ...   unsafe { to_u16_lanes(load16_unaligned(self.ptr.add(byte_index))) }
    |                             ^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `to_u16_lanes` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:137:18
    |
137 | ...   unsafe { to_u16_lanes(load16_unaligned(self.ptr.add(byte_index))) }
    |                ^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `simd_byte_swap` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:166:28
    |
166 |                     simd = simd_byte_swap(simd);
    |                            ^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `store8_unaligned` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:170:21
    |
170 | ...   store8_unaligned(other.as_mut_ptr().add(offset), simd);
    |       ^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `contains_surrogates` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:172:20
    |
172 |                 if contains_surrogates(simd) {
    |                    ^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `simd_byte_swap` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:262:25
    |
262 |                 first = simd_byte_swap(first);
    |                         ^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `simd_byte_swap` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:263:26
    |
263 |                 second = simd_byte_swap(second);
    |                          ^^^^^^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `simd_is_basic_latin` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:265:17
    |
265 |               if !simd_is_basic_latin(first | second) {
    |                   ^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `ascii_to_basic_latin`
    |
   ::: /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/ascii.rs:231:9
    |
231 | /         pub unsafe fn $name(
232 | |             src: *const $src_unit,
233 | |             dst: *mut $dst_unit,
234 | |             len: usize,
...   |
336 | |             None
337 | |         }
    | |_________- similarly named function `ascii_to_basic_latin` defined here

error[E0425]: cannot find function `simd_pack` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:268:26
    |
268 |             let packed = simd_pack(first, second);
    |                          ^^^^^^^^^ not found in this scope

error[E0425]: cannot find function `store16_unaligned` in this scope
   --> /home/tdynamos/.cargo/registry/src/index.crates.io-6f17d22bba15001f/encoding_rs-0.8.34/src/handles.rs:271:17
    |
271 |                 store16_unaligned(dst.as_mut_ptr().add(offset), packed);
    |                 ^^^^^^^^^^^^^^^^^ not found in this scope

Some errors have detailed explanations: E0412, E0425, E0432.
For more information about an error, try `rustc --explain E0412`.
error: could not compile `encoding_rs` (lib) due to 24 previous errors
warning: build failed, waiting for other jobs to finish...
💥 maturin failed===================>  ] 191/201: pyo3                        
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit status: 101": `env -u CARGO PYO3_ENVIRONMENT_SIGNATURE="cpython-3.11-64bit" PYO3_PYTHON="/tmp/build-env-c_wq7jrr/bin/python" PYTHON_SYS_EXECUTABLE="/tmp/build-env-c_wq7jrr/bin/python" "cargo" "rustc" "--features" "pyo3/extension-module" "--target" "armv7-linux-androideabi" "--message-format" "json-render-diagnostics" "--manifest-path" "/home/tdynamos/Expreiments/PythonTest/.buildozer/android/platform/build-armeabi-v7a/build/other_builds/pyreqwest_impersonate/armeabi-v7a__ndk_target_23/pyreqwest_impersonate/Cargo.toml" "--release" "--lib"`
Error: command ['maturin', 'pep517', 'build-wheel', '-i', '/tmp/build-env-c_wq7jrr/bin/python', '--compatibility', 'off'] returned non-zero exit status 1

ERROR Backend subprocess exited when trying to invoke build_wheel
```

--- a/pythonforandroid/recipes/pyreqwest_impersonate/__init__.py
+++ b/pythonforandroid/recipes/pyreqwest_impersonate/__init__.py
@@ -1,0 +1,33 @@
+from pythonforandroid.logger import info
+from pythonforandroid.recipe import RustCompiledComponentsRecipe
+
+
+class Pyreqwest_impersonateRecipe(RustCompiledComponentsRecipe):
+    version = "v0.4.5"
+    url = "https://github.com/deedy5/pyreqwest_impersonate/archive/refs/tags/{version}.tar.gz"
+
+    def get_recipe_env_post(self, arch, **kwargs):
+        env = super().get_recipe_env(arch, **kwargs)
+        env["ANDROID_NDK_HOME"] = self.ctx.ndk.llvm_prebuilt_dir
+        return env
+
+    def get_recipe_env_pre(self, arch, **kwargs):
+        env = super().get_recipe_env(arch, **kwargs)
+        env["ANDROID_NDK_HOME"] = self.ctx.ndk_dir
+        return env
+
+    def build_arch(self, arch):
+        # Why need of two env?
+        # Because there are two dependencies which accepts
+        # different ANDROID_NDK_HOME
+        self.get_recipe_env = self.get_recipe_env_pre
+        prebuild_ = super().build_arch
+        try:
+            prebuild_(arch)
+        except Exception:
+            info("pyreqwest_impersonate first build failed, as expected")
+            self.get_recipe_env = self.get_recipe_env_post
+            prebuild_(arch)
+
+
+recipe = Pyreqwest_impersonateRecipe()

